API: libpacemaker: Move crm_error functionality to libpacemaker

nrwahl2 opened the pull request:
Closes [T95](https://projects.clusterlabs.org/T95) and [T97](https://projects.clusterlabs.org/T97)

nrwahl2 commented on lib/common/output_text.c:
```diff
@@ -335,7 +335,9 @@ pcmk__formatted_vprintf(pcmk__output_t *out, const char *format, va_list args) {
     CRM_ASSERT(out != NULL);
 
-    len = vfprintf(out->dest, format, args);
+    if (pcmk__str_eq(out->fmt_name, "text", pcmk__str_none)) {
```
I did this for now as a quick fix. Ultimately we might want to add a manual print option to `pcmk__output_t`, so that we can call a member function in default formatters, instead of calling a function that's defined in `output_text.c` for text output objects.

If we don't give every output type its own manual print function, then we should probably move the functionality to a common file and guard it by output type like this.

clumens replied:
Adding a new `->print` function seems like a reasonable approach.  I can work on that really quick, unless you've already got something halfway done

clumens replied:
Looking at this again, there's very few places where this function is actually used in a default formatter.  It was really intended to be used in building text message functions and that's it.  I'd say that any other places it's being used is a bug and should be corrected there instead.  A manual printing function doesn't make sense for formatters other than text - what does printing to the screen look like for XML, or logs, or curses?

Aside from fixing those other usages, perhaps what we should do in pcmk__formatted_printf is return if fmt_name is "none", do the print if it's "text", and assert otherwise.

Note that everything here also applies to pcmk__indented_(v)printf.

clumens replied:
See #2848.

clumens commented on tools/crm_error.c:
```diff
@@ -78,11 +79,97 @@ get_strings(int rc, const char **name, const char **str)
+            pcmk__formatted_printf(out, "%s\n", ((name != NULL)? name : desc));
+        }
+        return pcmk_rc_ok;
+    }
```
You could just pass an extra bool argument to result_code_default to control whether to print this or not.  Various other format functions take flags for things like this.

nrwahl2 replied:
That's valid too. My thinking is that the bool would be ignored in the non-default (i.e., non-text) context, and that the `quiet` flag also conveys that it enables a compact form of the output.

nrwahl2 commented:
Rebased on current master and updated to incorporate the changes in the refactor PR #2838 that this depended on.

kgaillot commented on tools/crm_error.c:
```diff
@@ -57,11 +56,99 @@ static GOptionEntry entries[] = {
+        }
+        return out->info(out, "% *d: %-*s  %s", code_width, code, name_width,
+                         name, desc);
+    }
```
We don't have to calculate code_width and name_width every time, we can make them static variables

kgaillot commented on lib/pacemaker/pcmk_result_code.c:
```diff
@@ -0,0 +1,127 @@
+ * \param[in]     code   The result code
+ * \param[in]     type   Interpret \c code as this type of result code.
+ *                       Supported values: \c pcmk_result_legacy,
+ *                       \c pcmk_result_pcmk_rc, \c pcmk_result_crm_exitcode.
```
These names changed

kgaillot commented on lib/pacemaker/pcmk_result_code.c:
```diff
@@ -0,0 +1,127 @@
+ * \param[in,out] out    Output object
+ * \param[in]     type   The family of result codes to list. Supported
+ *                       values: \c pcmk_result_legacy, \c pcmk_result_pcmk_rc,
+ *                       \c pcmk_result_crm_exitcode.
```
also here

kgaillot commented on tools/Makefile.am:
```diff
@@ -71,7 +71,8 @@ crmadmin_LDADD		= $(top_builddir)/lib/pengine/libpe_status.la	\
 crm_error_SOURCES	= crm_error.c
-crm_error_LDADD		= $(top_builddir)/lib/common/libcrmcommon.la
+crm_error_LDADD		= $(top_builddir)/lib/common/libcrmcommon.la	\
+			  $(top_builddir)/lib/pacemaker/libpacemaker.la
```
other way around

kgaillot commented on include/pacemaker.h:
```diff
@@ -231,6 +231,21 @@ enum pcmk_rc_disp_flags {
+ * \param[in]     code   The result code
+ * \param[in]     type   Interpret \c code as this type of result code.
+ *                       Supported values: \c pcmk_result_legacy,
+ *                       \c pcmk_result_pcmk_rc, \c pcmk_result_crm_exitcode.
```
and here

kgaillot commented on cts/cts-cli.in:
```diff
@@ -538,6 +540,144 @@ function test_crm_mon() {
+    test_assert_validate $CRM_EX_OK 0
+
+    desc="Get multiple legacy return codes"
+    cmd="crm_error -- 1 2"
```
errno's and their descriptions vary by OS, so we should only test our custom codes.

That will be a problem for the full list, not sure of the best approach there. Maybe |head -something or |grep something to spot-check.

kgaillot reviewed (commented):
Nice, this feels much cleaner

nrwahl2 commented on tools/crm_error.c:
```diff
@@ -57,11 +56,99 @@ static GOptionEntry entries[] = {
+         * compact format for text output, or print nothing at all for none-type
+         * output.
+         */
+        if (pcmk__str_eq(out->fmt_name, "text", pcmk__str_none)) {
```
We only support `text`, `xml`, and `none`, so I think our options are:
* What I've got right now
* Have separate, probably mostly redundant formatters for `default` (`none`) and `text` to avoid this special handling
* Get rid of the `default` formatter altogether and convert this to a `text` formatter without the special check. I think you mentioned in a separate PR that, questions of style aside, if an output type doesn't have a formatter registered for a particular message then the message is just ignored.

kgaillot replied:
The last option sounds good

nrwahl2 commented:
Updates:
* Make `code_width` and `name_width` static variables to avoid redundant work.
* Reordered libraries in `crm_error_LDADD`
* Fixed enum value names in Doxygen comments
* Updated regression tests
  * Test list command for all three result code types, using `grep` to exclude system error codes
  * Don't use any system error codes in non-list tests
* Gave `text` its own output formatter function
* Gave `none` its own output formatter function that simply returns `pcmk_rc_no_output`.

This last bullet point is an option we haven't discussed yet, so I want to highlight it in particular (cc @clumens). I'm not tied to it. The idea is that if we support `none` output for a particular message, we don't want to log "Called unknown output message" even at a debug level, ideally. That makes it look like we didn't anticipate such a call.

However, we do want to make sure we print nothing for none-type. Up to now, we've relied on `out->info()` and `out->err()`, which are no-ops for none-type output, placed within a default formatter function that we share with `text` outputs. But sometimes for text outputs, we need to call `pcmk__formatted_printf()`, which would cause none-type outputs to print something. So in those cases, we want `text` to have its own designated formatter.

It feels wasteful to have a separate default formatter that makes the same calls as the text formatter minus `pcmk__formatted_printf()`, if we're going to use it only for none-type outputs that are expected to print nothing. And for formats besides `none`, `text`, and `xml` (if they're the only ones supported), it makes sense to log an "unknown output message" message.

Hence a formatter specifically for `none` that immediately returns `pcmk_rc_no_output`.

nrwahl2 commented:
FYI, CI test failure because `egrep: warning: egrep is obsolescent; using grep -E`. First I've heard...
